tools/reactive-controllers/src/RovingTabindex.ts

```diff
@@ -0,0 +1,304 @@
+    hostNoLongerContainsFocus(): void {
+        this.host.addEventListener('focusin', this.handleFocusin);
+        this.host.removeEventListener('focusout', this.handleFocusout);
+        this.host.removeEventListener('keydown', this.handleKeydown);
+        this.currentIndex = this.focusInIndex;
+        this.focused = false;
+        const currentElement = this.elements[this.currentIndex];
+        if (!currentElement) return;
+    }
+
+    isRelatedTargetAnElement(event: FocusEvent): boolean {
+        const relatedTarget = event.relatedTarget as null | Element;
+        return !this.elements.includes(relatedTarget as T);
+    }
+
+    handleFocusin = (event: FocusEvent): void => {
+        if (!this.isEventWithinListenerScope(event)) return;
+        if (this.isRelatedTargetAnElement(event)) {
+            this.hostContainsFocus();
+        }
+        const path = event.composedPath() as T[];
+        let targetIndex: number | undefined;
+        path.find((el) => {
+            targetIndex = this.elements.indexOf(el);
+            return targetIndex !== -1;
+        });
+        this.currentIndex = targetIndex ?? this.currentIndex;
+    };
+
+    handleFocusout = (event: FocusEvent): void => {
+        if (this.isRelatedTargetAnElement(event)) {
+            this.hostNoLongerContainsFocus();
+        }
+    };
+
+    acceptsEventCode(code: string): boolean {
+        if (code === 'End' || code === 'Home') {
+            return true;
+        }
+        switch (this.direction) {
+            case 'horizontal':
+                return code === 'ArrowLeft' || code === 'ArrowRight';
+            case 'vertical':
+                return code === 'ArrowUp' || code === 'ArrowDown';
+            case 'both':
+            case 'grid':
+                return code.startsWith('Arrow');
+        }
+    }
+
+    handleKeydown = (event: KeyboardEvent): void => {
+        if (!this.acceptsEventCode(event.code) || event.defaultPrevented) {
+            return;
+        }
+        let diff = 0;
+        switch (event.code) {
+            case 'ArrowRight':
+                diff += 1;
+                event.preventDefault();
+                break;
+            case 'ArrowDown':
+                diff += this.direction === 'grid' ? this.directionLength : 1;
+                event.preventDefault();
+                break;
+            case 'ArrowLeft':
+                diff -= 1;
+                event.preventDefault();
+                break;
+            case 'ArrowUp':
+                diff -= this.direction === 'grid' ? this.directionLength : 1;
+                event.preventDefault();
+                break;
+            case 'End':
+                this.currentIndex = 0;
+                diff -= 1;
+                event.preventDefault();
+                break;
+            case 'Home':
+                this.currentIndex = this.elements.length - 1;
+                diff += 1;
+                event.preventDefault();
+                break;
+        }
```

Comment:
Would this accomplish the same thing, or is there a reason not to bail if none of our codes match?

```suggestion
        switch (event.code) {
            case 'ArrowRight':
                diff += 1;
                break;
            case 'ArrowDown':
                diff += this.direction === 'grid' ? this.directionLength : 1;
                break;
            case 'ArrowLeft':
                diff -= 1;
                break;
            case 'ArrowUp':
                diff -= this.direction === 'grid' ? this.directionLength : 1;
                break;
            case 'End':
                this.currentIndex = 0;
                diff -= 1;
                break;
            case 'Home':
                this.currentIndex = this.elements.length - 1;
                diff += 1;
                break;
            default:
                return;
        }
        event.preventDefault();
```